### PR TITLE
Handle case when magento sale_order product is missing #6678

### PR DESCRIPTION
**Pull request by tarunbhardwaj**
As magento allow user to delete product even if it is used in
sale_order, it will throw error when connector try to fetch product from
magento server.

### CODE REVIEW

File: sale.py [hunk 1]
```diff
@@ -338,9 +364,8 @@ def get_item_line_data_using_magento_data(cls, order_data):
                     'unit': unit.id,
                     'quantity': Decimal(item['qty_ordered']),
                     'note': item['product_options'],
-                    'product': ProductTemplate.find_or_create_using_magento_id(
-                        item['product_id'],
-                    ).products[0].id
+                    'product': product,
+                    'has_magento_exception': not product,
```
Comment (sharoonthomas) [inline sale.py]: (product is None)

File: sale.py [hunk 2]
```diff
@@ -204,6 +205,21 @@ def check_store_view_instance(self):
         return True
 
     @classmethod
+    def confirm(cls, sales):
+        "Validate sale before confirming"
+        SaleLine = Pool().get('sale.line')
+
+        for sale in sales:
+            has_exception = SaleLine.search([
```
Comment (sharoonthomas) [inline sale.py]: `if sale.has_magento_exception:`

File: sale.py [hunk 3]
```diff
@@ -331,6 +347,16 @@ def get_item_line_data_using_magento_data(cls, order_data):
         for item in order_data['items']:
             if not item['parent_item_id']:
                 # If its a top level product, create it
+                try:
+                    product = ProductTemplate.find_or_create_using_magento_id(
+                        item['product_id'],
+                    ).products[0]
+                except xmlrpclib.Fault, exception:
+                    if exception.faultCode == 101:
```
Comment (prakashpp) [inline sale.py]: why not store exception message to notes?

File: magento_.py [hunk 4]
```diff
@@ -1182,3 +1182,25 @@ def do_export_(self, action):
...
+class MagentoException(ModelSQL, ModelView):
+    """
+    Magento Exception model
+    """
+    __name__ = 'magento.exception'
+
+    origin = fields.Reference(
+        "Origin", selection='models_get', select=True,
+    )
+    exception_log = fields.Text('Exception Log')
```
Comment (prakashpp) [inline magento_.py]: log only. `sale.line.exception.log`

**Pull request merged**